src/lib.rs

```diff
@@ -105,3 +105,124 @@ mod url_tests {
         assert!(result.is_ok());
     }
 }
+
+#[cfg(test)]
+mod management_tests {
+    use crate::{
+        management::list_files::{ListFiles, Operator, Options, SearchQuery, SearchQueryBuilder},
+        ImageKit,
+    };
+
+    #[tokio::test]
+    async fn list_files() {
+        let imagekit = ImageKit::from_env().unwrap();
+        let result = imagekit.list_files(Options::default()).await;
+        assert!(result.is_ok());
+    }
+
+    #[tokio::test]
+    async fn list_files_with_raw_query_string_search() {
+        let imagekit = ImageKit::from_env().unwrap();
+        let query = SearchQueryBuilder::raw_query_string("name=\"default-image.jpg\"").build();
+        let options = Options::new().search_query(query);
+        let result = imagekit.list_files(options).await;
+        assert!(result.is_ok());
+    }
+
+    #[tokio::test]
+    async fn list_files_without_builder() {
+        let imagekit = ImageKit::from_env().unwrap();
+        let query = SearchQuery("name=\"default-image.jpg\"".into());
+        let options = Options::new().search_query(query);
+        let result = imagekit.list_files(options).await;
+        assert!(result.is_ok());
+    }
+
+    #[tokio::test]
+    async fn list_files_with_name_search() {
+        let imagekit = ImageKit::from_env().unwrap();
+        let query = SearchQueryBuilder::name(Operator::EqualTo, "\"default-image.jpg\"").build();
```

Comment:
Yes, operator is a kind of unusual. But, is very common in a rule engine world. 

If done in the style of name_equals, we will need to have 105 methods to cater to the available  combinations of operators & field names. I think the effort needed to write so many methods is not worth it. Any future addition of a single field will also add a lot of additional methods in the builder. I'm nto sure if we really need this approach.

One downside of operator vs name_equals approach is: not all combinations are available for all fields with all operators. However, if an unavailable combination is given as input, imagekit will anyway reject the request. So, it is probably a little 'nicer', not really a 'required' thing.

Also, the operator approach makes it easier if someone wants to build a UI with user definable filters more easier. For example, we could have an UI where the user can pick the field name, operator and the value as 3 different inputs fields. (Of course, certain combinations of names/operators are not allowed, which may need ui level validations, or, it may generate error from imagekit end. Again, probably a nice to do. Not very required.)

In my personal opinion, I am not sure many people will really make use of the builder, and we can spend more effort on it, if there is a real demand. (I personally will be using the query string directly (if possible as a raw string), as it is a lot more convenient to check using the rest endpoint, and eventually copy-paste the query string to the rust code) 

:)

Reply:
Interesting. Usually you use a builder when a struct has some many fields to avoid too complex constructors. Buliders are very common in Rust code as it allows you to write easy to read code and also clever validation process.

I see [here that theres not to many parameters](https://docs.imagekit.io/api-reference/media-api/list-and-search-files#supported-parameters), is more about their combinations and about _which parameter supports which field_.

I think, you could implement support for what you need as of now? I dont see any `raw` string support, but instead, metadata with dynamic keys.

This can be written like this as well:

```rust
const NAME_CONDITIONS: [Condition, 1] = [Condition::Equal];

enum Condition {
  Equals
}

#[derive(Default)]
struct SearchQuery {
  name: Option<(String, Operator)>,
  // -- other fields --
}

impl SearchQuery {
  pub fn new() -> Self {
    SearchQuery::default()
  }

  pub fn name<T: Into<String>>(&mut self, v: T, cond: Condition) -> Result<Self, SearchQueryError> {
    if NAME_CONDITIONS.contains(cond) {
      self.name = Some((v.into(), cond));
      return Ok(self);
    }

    Err(SearchQueryError::UnsupportedCondition("name", cond))
  }
}
```

We dont have to implement every parameter at once, we can add more later with this approach. But with thiswe can ensure bad code dont get to production. If a user creates an invalid query usnig a string it will take time to find the issue.

WDYT?

Reply:
Basically you wanted to rename "Operator" to "Condition"? I thought operator is a more commonly understood term for programmers. I'm just referring here.. https://doc.rust-lang.org/book/appendix-02-operators.html

And, I do like the idea of adding additional checks to see if the given combination is indeed valid or. And yes, I agree that, this is nice to do feature on a field by field basis, and can be done at later point too.

Reply:
> #[derive(Default)]
> struct SearchQuery {
>   name: Option<(String, Operator)>,
>   // -- other fields --
> }

As far as representing SearchQuery in this manner is concerned, I like it, and I would probably prefer it, if it as simple as that. However, as I mentioned few days back, SearchQuery is NOT a simple object, and it can not be represented by a struct easily, at least as far as I can think of.

For example, please try representing "private = true and (size > 200 or (tags IN [\"summer-sale\"]))" in a struct. This string is actually a "SearchQuery". But, can we represent this as a struct? I highly doubt it. If we can't faithfully represent a struct to its real world equivalent, it is better to simplify its representation, than trying to represent it half way through, which will complicate the design / coding, just for no tangible benefit.

Reply:
> > #[derive(Default)]
> > struct SearchQuery {
> > name: Option<(String, Operator)>,
> > // -- other fields --
> > }
> 
> As far as representing SearchQuery in this manner is concerned, I like it, and I would probably prefer it, if it as simple as that. However, as I mentioned few days back, SearchQuery is NOT a simple object, and it can not be represented by a struct easily, at least as far as I can think of.
> 
> For example, please try representing "private = true and (size > 200 or (tags IN ["summer-sale"]))" in a struct. This string is actually a "SearchQuery". But, can we represent this as a struct? I highly doubt it. If we can't faithfully represent a struct to its real world equivalent, it is better to simplify its representation, than trying to represent it half way through, which will complicate the design / coding, just for no tangible benefit.

May be this representation is possible, if we do something similar to a 'Value' in serde, or how they have done it in https://docs.rs/rule/latest/rule/. (in such casees, I guess, it will kind of get into some sort of a recursion/self reference, etc). But, I was thinking should we get into so much complication, at this point. May be this struct can be improved later too; and it can be done without any impact to users, as it is an internal implementation to the library only.